Set tight permissions for key files

Lekensteyn opened the pull request:
Ensure that key files are not world-readable.

temp_umask is based on test.support.temp_umask.

Fixes #29

---

If older Python versions do not have to be supported, then the test code could become something like:

``` py
@temp_umask(0o022)
def test_it():
    ...
```

Lekensteyn commented on simp_le.py:
```diff
@@ -371,7 +396,8 @@ def load(self):  # pylint: disable=missing-docstring
         output = (self.dump_cert(chain_cert) for chain_cert in data.chain)
-        with open(self.path, 'wb') as chain_file:
+        open_func = open_keyfile if data.key else open
+        with open_func(self.path, 'wb') as chain_file:
```
Ok, apparently a ChainFile behaves likes its name, it does not store the key even though the FullChainFile implementation suggests otherwise. So this part is not needed.

This becomes different with a change such as #33, should it always set the permission to 0600 for security reasons? If the user requires weaker permissions, the file can be touched before creating (or chmod can be invoked afterwards).

danmilon commented:
World readable private key? ouch!
LGTM.

Lekensteyn committed:
Set tight permissions for key files

Ensure that key files are not world-readable.

temp_umask is based on test.support.temp_umask.

Fixes #29

Lekensteyn commented:
Updated patchset against latest master, added a rough separate test case as AccountKey got removed.

kuba commented on simp_le.py:
```diff
@@ -1183,5 +1210,33 @@ def test_error_exit_codes(self, dummy_stderr):
+    def setUp(self):
+        self.root = tempfile.mkdtemp()
+        self.key_path = os.path.join(self.root, 'key.pem')
+        test_key = os.path.join('testdata', "rsa2048_key.pem")
```
Please use `gen_pkey`: I'd like all tests to be self-contained within one script (e.g. so that they can be run from binaries created with PyInstaller), so we would have to inline entire key, which doesn't look great. If at some point that becomes of an performance concern, we will think about it then.

BTW if we were to use testdata, we would use `pkg_resources` to load such.

Lekensteyn replied:
I've tried that, but that seems to require so many indirections (implementation details) that I decided to use this approach instead. I have a small preference to fixed files for reproducibility and faster key generation.

Also, the current `IOPlugin.Data` thing does not scale, what about setting defaults to None? Something like (out of my head):

``` py
IOPlugin.Data.__defaults__ = (None, ) * 4
```

Then it should become something like `key_data = IOPlugin.Data(key=gen_pkey(1024))` (+ some indirections, I could not get this to work).

kuba commented on simp_le.py:
```diff
@@ -78,6 +82,26 @@ class TestCase(unittest.TestCase):
+    # Windows-only
+    if 'b' in mode and hasattr(os, "O_BINARY"):
+        flags |= os.O_BINARY  # pylint: disable=no-member
+    fild = os.open(filename, flags, 0o600)
```
- inline `filed`
- I think we want `0o660`

Lekensteyn replied:
inlining, ok.

600 should be the default, at least postgres (or was it just the node-pg or PHP library?) require such strict permissions. `ssh-keygen` also creates such files. And for simplicity reasons, I would not add another option to change it.

Rationale: you can chmod afterwards, or create an empty file before that with appropriate permissions. (Actually, empty files seem to trigger `load_from_content("")`, maybe that should also return `EMPTY_DATA`?)

kuba commented on simp_le.py:
```diff
@@ -78,6 +82,26 @@ class TestCase(unittest.TestCase):
+        os.umask(oldmask)
+
+
+def open_keyfile(filename, mode):
```
that's not specific to keys, and is used generically elsewhere when `sensitive=True`, so please rename to `open_sensitive`

kuba commented on simp_le.py:
```diff
@@ -1183,5 +1210,33 @@ def test_error_exit_codes(self, dummy_stderr):
                 assert False
 
 
+class IOPluginTests(TestCase):
```
This should be `KeyFileTest` and by that I mean both naming and implementation: directly initialize `KeyFile` instead of using `IOPlugin.registered['key.pem']`

Lekensteyn replied:
This was actually a deliberate choice, the user interface is "key.pem", so if for some reason this was changed, it should still work (as the user expects). I'll change it anyway while making it specific for KeyFilePlugin (originally I intended to check account_key too).

kuba commented on simp_le.py:
```diff
@@ -1183,5 +1210,33 @@ def test_error_exit_codes(self, dummy_stderr):
+        os.remove(self.key_path)
+        with temp_umask(0o022):
+            keyfile.save(key_data)
+            file_mode = os.stat(self.key_path).st_mode
```
inline `file_mode`

kuba commented on simp_le.py:
```diff
@@ -1183,5 +1210,33 @@ def test_error_exit_codes(self, dummy_stderr):
+        with temp_umask(0o022):
+            keyfile.save(key_data)
+            file_mode = os.stat(self.key_path).st_mode
+            self.assertEqual(file_mode & stat.S_IROTH, 0)
```
this stmt can be move out of the context manager block

kuba commented on simp_le.py:
```diff
@@ -1183,5 +1210,33 @@ def test_error_exit_codes(self, dummy_stderr):
+
+        key_data = keyfile.load()
+        self.assertNotEqual(key_data, IOPlugin.EMPTY_DATA)
+        os.remove(self.key_path)
```
if you do tests other way round: first gen, then save, then load, you don't have to do that `os.remove` as it's cleanup in `tearDown`

Lekensteyn replied:
Yes, this was a workaround for my inability to use `gen_pkey`. I am open for suggestions :)

kuba commented:
There are 3 parties potentially involved with the typical usage scenario:
1. user under which the client is run
2. user that runs httpd
3. root

If we assume that those three parties are different users (say `nobody`, `nginx`, `root`), then there are some usability problems, because:
- (1) needs to open and write: account key, chain, cert and key
- (2) needs to read chain, cert and key
- we don't run as `root`, so we cannot `chmod` or `chown`

Restricting key file to `0o600` would mean that httpd is not able to read necessary files :(

I would like those kind of cases to be clearly sorted out (and by that I mean short design doc, possibly included in the repo itself) before merging.

danmilon commented:
AFAIK apache runs as root to bind to privilaged ports and/or read the private key, and then spawns other processes as a non-privilaged user (e.g. `www-data`) to do the actual request handling.

I don't know if other web servers work similarly.

Lekensteyn committed:
Add tests for keeping keyfiles secure

It is pretty important for keyfiles to be kept secret. Add a test that
will catch possible changes that violate this.

Lekensteyn committed:
Simplify IOPlugin.Data initialization

Remove the need to specify every component of Data. Note that after this
change, the IOPlugin.persisted interface may return None as falsive
value. Since there is no identity check (x is False), it is fine.

Also remove account_key and key from FullChainFile, it is not stored by
ChainFile.

Lekensteyn committed:
Ignore existing but empty files

Lekensteyn committed:
Add note on file permissions

Lekensteyn commented:
Comments have been addressed, the code has also been made simpler (both the test and IOPlugin.Data values).

kuba commented on testdata/rsa2048_key.pem:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
you forgot to remove this file

kuba commented on simp_le.py:
```diff
@@ -243,8 +266,9 @@ class IOPlugin(object):
     - for `cert`: certificate, an instance of `OpenSSL.crypto.X509`
     - for `chain`: certificate chain, a list of `OpenSSL.crypto.X509` instances
     """
+    Data.__new__.__defaults__ = (None,) * 4
```
I have never ever seen this. Any docs?

kuba replied:
Also, please keep PRs scope limited. This kind of change should go into a different PR.

Lekensteyn replied:
I found this approach on SO, here is one for example: 
https://stackoverflow.com/questions/11351032/named-tuple-and-optional-keyword-arguments

The  [Py2 Data model](https://docs.python.org/2/reference/datamodel.html) and [Py3 Data model](https://docs.python.org/3/reference/datamodel.html) docs allow this:

> **Attribute**: `__defaults__`  
> **Meaning**: A tuple containing default argument values for those arguments that have defaults, or None if no arguments have a default value  
> Writable

kuba commented on README.rst:
```diff
@@ -71,6 +71,21 @@ Installation
     . venv/bin/activate
 
+File permissions
+----------------
```
missing newline after header, rst might not render correctly

kuba commented on simp_le.py:
```diff
@@ -318,6 +342,10 @@ def load(self):
             raise
+        if not content:
+            # Empty files may exist when their parent directory is not
+            # writable, or when the file mode needs to be overridden.
```
I'm not sure what's going on here. Either inline comment needs to be improved, or something is wrong.

kuba commented on simp_le.py:
```diff
@@ -848,8 +874,7 @@ def test(args):
     """Do plugins cover all components (key/cert/chain)?"""
-    persisted = IOPlugin.Data(
-        account_key=False, key=False, cert=False, chain=False)
+    persisted = IOPlugin.Data()
```
That's clearly wrong, as `IOPlugin.Data()` would contain `None` instead of `False`

kuba commented on README.rst:
```diff
@@ -71,6 +71,21 @@ Installation
 
+File permissions
+----------------
+Newly created key files are only accessible by ``simp_le`` user (mode
```
"user under which `simp_le` runs" - otherwise it's a bit confusing

kuba commented on README.rst:
```diff
@@ -71,6 +71,21 @@ Installation
+600) while certificates are restricted by the umask setting.
+
+Should you need more relaxed permissions, either create an empty file
+first or modify the permission afterwards. ``simp_le`` will not modify
```
I don't see any tests for this behavior (one for empty file, one for persisted mode). Also, I'm not sure If like like the "empty file" solution...

kuba commented:
When I said "design doc" I expected discussion about alternative solutions, giving it's pros and cons. Since you picked one particular implementation, it's rather hard to discuss it. I'm not convinced that the solution presented in this PR is the best one.  Notably, that official client has similar problems, and AFAIK there might not be perfect solution to the problem - but at least I would like to know how much we are limited by chosen solution.

Please keep PRs in their original scope. This is not the right place to discuss `Data.__new__.__defaults__` magic. Please revert and feel free to submit a separate PR.

kuba commented:
For example, why not simply allow `--sensitive_files_mode 660`?
